### PR TITLE
Adding Spatial Transformers w/CuDNN support

### PR DESCRIPTION
**Pull request by soumith**
TODO:
- [x] add cudnn bindings for Transformer functions
- [x] add CPU functions in THNN
- [x] unit tests
- [x] documentation

Fixes https://github.com/pytorch/pytorch/issues/496

### CODE REVIEW

**Comment by soumith** [comment]
This is ready for review / good to go.
The next thing that's left for SpatialTransformers is to add the grid generator from an affine 6D matrix. But this PR itself is hugely useful to various vision folks, for example @jcjohnson who uses the bilinear resampling

**Comment by soumith** [comment]
I've added the Affine grid generator as well. Once this is merged, the SpatialTransformers issue can be closed.

File: torch/csrc/cudnn/AffineGridGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+void setSamplerDescriptor(SpatialTransformerDescriptor& desc,
+			  cudnnDataType_t dataType,
+			  int N, int C, int H, int W)
+{
+  int inputSize[4] = {N, C, H, W};
+  desc.set(dataType, 4, inputSize);
+}
+
+void* tensorPointer(cudnnDataType_t dataType, THVoidTensor* tensor)
```
Comment (albanD) [inline torch/csrc/cudnn/AffineGridGenerator.cpp]: This function is duplicated in `GridSampler.cpp`, could this be avoided?

Comment (albanD) [inline torch/csrc/cudnn/AffineGridGenerator.cpp]: This function is duplicated in `GridSampler.cpp`, could this be avoided?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -821,6 +822,55 @@ def upsample_bilinear(input, size=None, scale_factor=None):
...
+
+
+def affine_grid(theta, size):
+    """Generates a 2d flow field, given a batch of affine matrices :attr:`theta`
+    Generally used in conjunction with :function:`grid_sample` to
+    implement Spatial Transformer Networks.
+
+    Args:
+        theta (Variable): input batch of affine matrices (N x 2 x 3)
+        size (torch.Size): the target output image size (N x H x W x 2)
```
Comment (albanD) [inline torch/nn/functional.py]: This looks slightly confusing we should say explicitly that we need to give `(N x C x H x W)` in order to get an output of size `(N x H x W x 2)`.

Also I see that the C parameter is used in the `cudnn` function, but what is it used for? It should be independent of the value of `C` no ?

File: torch/lib/THNN/generic/SpatialGridSamplerBilinear.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	  real se_val = SAFE_GET(input, ix_se, iy_se, n, c, H, W);
+	  real out_val = nw_val * nw + ne_val * ne + sw_val * sw + se_val * se;
+	  THTensor_fastSet4d(output, n, c, h, w, out_val);
+	}
+      }
+    }
+  }
+}
+
+#define SAFE_SET(input, x, y, n, c, H, W, value)		\
```
Comment (albanD) [inline torch/lib/THNN/generic/SpatialGridSamplerBilinear.c]: shouldn't this be called `SAFE_ADD` ?

File: torch/lib/THNN/generic/SpatialGridSamplerBilinear.c [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	  giy -= ne_val * (ix - ix_sw) * gradout;
+	  giy += sw_val * (ix_ne - ix) * gradout;
+	  giy += se_val * (ix - ix_nw) * gradout;
+	}
+
+	// un-normalize gradGrid values back to [-1, 1] constraints
+	gix = gix * (W - 1) / 2;
+	giy = giy * (H - 1) / 2;
+
+#pragma omp critical(SpatialGridSamplerBilinear_updateGridGrad)
```
Comment (albanD) [inline torch/lib/THNN/generic/SpatialGridSamplerBilinear.c]: Why is this a critical section? we are out of the loop on `c` so each thread will access a different elements of `gradGrid` no (I am far from comfortable with OpenMP so I am most certainly missing something)?

**Review by albanD — COMMENTED** [review]
I see that the `cudnn` binding convert all sizes and strides to `int`. Is that because the cudnn functions only accept `int`s and not `long`s ?

File: torch/nn/functional.py (hunk 2 above)
Comment (soumith) [inline torch/nn/functional.py]: The C parameter is used in cudnn's SpatialTransformerDescriptor. It's not needed (for example, CPU doesn't use it), but that's how CuDNN API seems to be.

I'll fix the confusing wording.

File: torch/lib/THNN/generic/SpatialGridSamplerBilinear.c (hunk 3 above)
Comment (soumith) [inline torch/lib/THNN/generic/SpatialGridSamplerBilinear.c]: sure, will change.

File: torch/lib/THNN/generic/SpatialGridSamplerBilinear.c (hunk 4 above)
Comment (soumith) [inline torch/lib/THNN/generic/SpatialGridSamplerBilinear.c]: yea you're right. I'll remove the parallelism on `c` and independently i can remove this critical section.

**Comment by soumith** [comment]
@albanD yes cudnn only supports int sizes and strides, so we often do this conversion. I presume because long indexing math is much more expensive on GPU which doesn't have as many double-width cores, they made this tradeoff.

File: torch/csrc/cudnn/AffineGridGenerator.cpp (hunk 1 above)
Comment (ngimel) [inline torch/csrc/cudnn/AffineGridGenerator.cpp]: It is also duplicated in BatchNorm.cpp, and Conv.cpp has a slightly different flavor of the same function.

**Comment by edgarriba** [comment]
@soumith tried on my local tests with images and seems to work. Thanks!

File: torch/csrc/cudnn/AffineGridGenerator.cpp (hunk 1 above)
Comment (csxmli2016) [inline torch/csrc/cudnn/AffineGridGenerator.cpp]: Do you have SpatialGridSamplerBilinear.cu in cuda

Comment (soumith) [inline torch/csrc/cudnn/AffineGridGenerator.cpp]: no, for the CUDA side we use CuDNN

**Comment by leao1995** [comment]
@soumith I want to perform grid sample on volumetric data, but current function only support 4D tensors. Is there a way to realize this? Or I have to create one?

**Comment by soumith** [comment]
@leao1995 unfortunately you will have to create one.
